File: elixir/lib/2024/03_mull_it_over/aoc202403.ex
defmodule AOC2024.Day03 do
  @moduledoc """
  Advent of Code 2024, day 3: Mull it Over.
  """
  require AOC

  @doc """
  Parse input.
  """
  def parse(puzzle_input) do
    puzzle_input |> String.split("\n", trim: true) |> Enum.join(" ") |> do_or_dont()
  end

  @doc """
  Split a text on "don't()" and "do()"

  ## Example

      iex> do_or_dont("First don't() do what you want then do() it anyway")
      ["First ", " do what you want then ", " it anyway"]
  """
  def do_or_dont(text), do: do_or_dont(text, [], true)
  def do_or_dont("", parts, _), do: parts |> Enum.reverse()

  def do_or_dont(text, parts, true) do
    case String.split(text, "don't()", parts: 2) do
      [left, right] -> do_or_dont(right, [left | parts], false)
      [rest] -> do_or_dont("", [rest | parts], false)
    end
  end

  def do_or_dont(text, parts, false) do
    case String.split(text, "do()", parts: 2) do
      [left, right] -> do_or_dont(right, [left | parts], true)
      [rest] -> do_or_dont("", [rest | parts], true)
    end
  end

  @doc """
  Solve part 1.
  """
  def part1(data) do
    data |> Enum.join("") |> find_multiplies() |> Enum.sum()
  end

  @doc """
  Solve part 2.
  """
  def part2(data) do
    data |> Enum.take_every(2) |> Enum.join("") |> find_multiplies() |> Enum.sum()
  end

  @doc """
  Find occurences of "mul(num1,num2)" in text and carry out the multiplications.

  ## Example

      iex> find_multiplies("First mul(4,2) then mul(44,46) and end with mul")
      [8, 2024]
  """
  def find_multiplies(text) do
    Regex.scan(~r/mul\((\d{1,3}),(\d{1,3})\)/, text, capture: :all_but_first)
    |> Enum.map(fn [first, second] -> String.to_integer(first) * String.to_integer(second) end)
  end

  def main(args) do
    Enum.map(args, fn path -> AOC.solve(path, &parse/1, &part1/1, &part2/1) end)
  end
end

File: python/src/2024/03_mull_it_over/aoc202403.py
"""AoC 3, 2024: Mull It Over."""

# Standard library imports
import pathlib
import sys

# Third party imports
import parse

MUL = parse.compile("mul({first:3d},{second:3d})")


def parse_data(puzzle_input):
    """Parse input. Break lines on occurences of do()"""
    return puzzle_input.replace("\n", "").split("do()")


def part1(data):
    """Solve part 1."""
    return sum(multiplies("".join(data)))


def part2(data):
    """Solve part 2."""
    return sum(multiplies("".join(line.split("don't")[0] for line in data)))


def multiplies(text):
    """Find multiply instructions in text and do the multiplication"""
    return (mul["first"] * mul["second"] for mul in MUL.findall(text))


def solve(puzzle_input):
    """Solve the puzzle for the given input."""
    data = parse_data(puzzle_input)
    yield part1(data)
    yield part2(data)


if __name__ == "__main__":
    for path in sys.argv[1:]:
        print(f"\n{path}:")
        solutions = solve(puzzle_input=pathlib.Path(path).read_text().rstrip())
        print("\n".join(str(solution) for solution in solutions))
